File: julia/2015/day06b/day06b.jl
#!/usr/bin/env julia

struct Position
  x :: Int32
  y :: Int32
end

struct Bounds
  upperLeft :: Position
  lowerRight :: Position
end

function usage()
    progname = PROGRAM_FILE
    println(stderr, "usage: $progname <input file>")
    exit(1)
end

function perform(grid::Array{Int32,2}, action::String, bounds::Bounds)
    for row in bounds.upperLeft.y:bounds.lowerRight.y
        for col in bounds.upperLeft.x:bounds.lowerRight.x
            if action == "turn on"
                grid[row, col] += 1
            elseif action == "turn off"
                grid[row, col] = (grid[row, col] > 0) ? (grid[row, col] - 1) : 0
            elseif action == "toggle"
                grid[row, col] += 2
            end
        end
    end
end

function process(content::String)::Int32
    grid = zeros(Int32, 1000, 1000)
    for line in split(content, '\n', keepempty=false)
        matches = match(r"(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)", line)
        if matches != nothing
            action :: String = matches.captures[1]
            bounds = Bounds(Position(parse(Int32, matches.captures[2]) + 1,
                                     parse(Int32, matches.captures[3]) + 1),
                            Position(parse(Int32, matches.captures[4]) + 1,
                                     parse(Int32, matches.captures[5]) + 1))
            perform(grid, action, bounds)
        end
    end
    return sum(grid)
end

function main()
    args = ARGS
    if length(args) < 1
        usage()
    end
    filename = args[1]
    try
        content = read(filename, String)
        result = process(content)
        println("result = $result")
    catch e
        println(stderr, "error reading file: $e")
        exit(1)
    end
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end

File: dart/2015/day06b/day06b.dart
#!/usr/bin/env dart

import 'dart:io';

const int MAX_ROW = 1000;
const int MAX_COL = 1000;

void usage(String progName) {
    print('usage: $progName <input file>');
    exit(1);
}

void perform(List<int> grid, String action, int r1, int c1, int r2, int c2) {
    for ( var row = r1; row <= r2; row++ ) {
        for ( var col = c1; col <= c2; col++ ) {
            var index = row * MAX_COL + col;
            switch ( action ) {
            case 'turn on':
                grid[index] += 1;
                break;
            case 'turn off':
                grid[index] = (grid[index] > 0) ? grid[index] - 1 : 0;
                break;
            case 'toggle':
                grid[index] += 2;
                break;
            }
        }
    }
}

int sum(List<int> grid) {
    var total = 0;
    for ( var row = 0; row < MAX_ROW; row++ ) {
        for ( var col = 0; col < MAX_COL; col++ ) {
            var index = row * MAX_COL + col;
            total += grid[index];
        }
    }
    return total;
}

int process(String filename) {
    var grid = List<int>.filled(MAX_ROW * MAX_COL, 0);
    final lines = File(filename).readAsLinesSync();
    for ( final line in lines ) {
        var re = RegExp(r'(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)');
        var match = re.firstMatch(line);
        if ( match != null ) {
            var action = match.group(1)!;
            var r1 = int.parse(match.group(2)!);
            var c1 = int.parse(match.group(3)!);
            var r2 = int.parse(match.group(4)!);
            var c2 = int.parse(match.group(5)!);
            perform(grid, action, r1, c1, r2, c2);
        }
    }
    return sum(grid);
}

void main(List<String> args) {
    if ( args.isEmpty ) {
        usage(Platform.script.pathSegments.last);
    }
    final filename = args[0];
    final result = process(filename);
    print('result = $result');
}
